zerver/views/registration.py

```diff
@@ -169,10 +170,12 @@ def accounts_register(request):
         if 'timezone' in request.POST and request.POST['timezone'] in get_all_timezones():
             timezone = request.POST['timezone']
 
-        try:
-            existing_user_profile = get_user_profile_by_email(email)
-        except UserProfile.DoesNotExist:
-            existing_user_profile = None
+        existing_user_profile = None
+        if not realm_creation:
+            try:
+                existing_user_profile = get_user(email, realm)
+            except UserProfile.DoesNotExist:
+                pass
```

Comment:
This hunk is in the commit "auth: Replace user_email_is_unique validator.", but it doesn't feel like it fits that description. Perhaps it should be its own commit?

Also the logic is a little confusing to follow -- I think it'd be clearer with an if/else, and three mutually exclusive and exhaustive assignments to `existing_user_profile`, rather than flowing through.

Reply:
Yeah, I'd extract this as a commit on the bottom, and make it more readable.